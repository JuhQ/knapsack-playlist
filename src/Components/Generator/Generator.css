button.preset {
  margin-right: 3px;
}

button.preset.selected {
  background: #f96dce;
}
