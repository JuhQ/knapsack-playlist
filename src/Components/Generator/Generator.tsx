import React, { useState } from "react";
import { Container, Form, Input } from "semantic-ui-react";

interface Props {
  onSubmit: (value: { dataSetSize: number; playlistLength: number }) => void;
}

const Generator: React.FC<Props> = ({ onSubmit }: Props) => {
  const [dataSetSize, setDataSetSize] = useState<number>(20);
  const [playlistLength, setPlaylistLength] = useState<number>(3600);

  return (
    <Container>
      <Form
        onSubmit={() => {
          onSubmit({ dataSetSize, playlistLength });
        }}
      >
        <div>
          Size of data set:
          <br />
          <small>
            This is the amount of random items to be selected from the database
            for the knapsack algorithm
          </small>
          <br />
          <Input
            size="large"
            type="number"
            label={{ basic: true, content: "videos" }}
            labelPosition="right"
            placeholder="Enter desired data set size"
            value={dataSetSize}
            onChange={({ target }) => {
              setDataSetSize(Number(target.value));
            }}
          />
        </div>
        <div>
          Length of playlist:
          <br />
          <Input
            size="large"
            type="number"
            label={{ basic: true, content: "seconds" }}
            labelPosition="right"
            placeholder="Enter desired playlist length"
            value={playlistLength}
            onChange={({ target }) => {
              setPlaylistLength(Number(target.value));
            }}
          />
        </div>
        <button type="submit">Submit</button>
      </Form>
    </Container>
  );
};

export default Generator;
