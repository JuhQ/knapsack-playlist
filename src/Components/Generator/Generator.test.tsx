import { shallow } from "enzyme";
import React from "react";
import renderer from "react-test-renderer";

import Generator from "./Generator";

describe("Generator", () => {
  it("should render", () => {
    const component = renderer.create(
      <Generator
        totalDataset={1200}
        totalLength={100000}
        onSubmit={() => null}
      />
    );
    const tree = component.toJSON();
    expect(tree).toMatchSnapshot();
  });

  it("should be able to call onSubmit function", () => {
    const handleSubmit = jest.fn();
    const wrapper = shallow(
      <Generator
        totalDataset={1200}
        totalLength={100000}
        onSubmit={handleSubmit}
      />
    );
    expect(handleSubmit).not.toHaveBeenCalled();
    wrapper.find("Form").simulate("submit");
    expect(handleSubmit).toHaveBeenCalled();
    expect(handleSubmit).toHaveBeenCalledWith({
      dataSetSize: 20,
      playlistLength: 3600,
    });
  });

  it("should be able to call onSubmit function with changed values", () => {
    const handleSubmit = jest.fn();
    const wrapper = shallow(
      <Generator
        totalDataset={1200}
        totalLength={100000}
        onSubmit={handleSubmit}
      />
    );
    expect(handleSubmit).not.toHaveBeenCalled();

    wrapper
      .find("Input")
      .first()
      .simulate("change", { target: { value: 200 } });
    wrapper
      .find("Input")
      .last()
      .simulate("change", { target: { value: 36000 } });

    wrapper.find("Form").simulate("submit");

    expect(handleSubmit).toHaveBeenCalled();
    expect(handleSubmit).toHaveBeenCalledWith({
      dataSetSize: 200,
      playlistLength: 36000,
    });
  });

  describe("inputs", () => {
    it("should be able to call onChange function on the dataset size input", () => {
      const handleSubmit = jest.fn();
      const wrapper = shallow(
        <Generator
          totalDataset={1200}
          totalLength={100000}
          onSubmit={handleSubmit}
        />
      );
      expect(handleSubmit).not.toHaveBeenCalled();
      wrapper
        .find("Input")
        .first()
        .simulate("change", { target: { value: 200 } });
      wrapper.find("Form").simulate("submit");

      expect(handleSubmit).toHaveBeenCalled();
      expect(handleSubmit).toHaveBeenCalledWith({
        dataSetSize: 200,
        playlistLength: 3600,
      });
    });

    it("should be able to call onChange function on the playlist length input", () => {
      const handleSubmit = jest.fn();
      const wrapper = shallow(
        <Generator
          totalDataset={1200}
          totalLength={100000}
          onSubmit={handleSubmit}
        />
      );
      expect(handleSubmit).not.toHaveBeenCalled();
      wrapper
        .find("Input")
        .last()
        .simulate("change", { target: { value: 200 } });
      wrapper.find("Form").simulate("submit");

      expect(handleSubmit).toHaveBeenCalled();
      expect(handleSubmit).toHaveBeenCalledWith({
        dataSetSize: 20,
        playlistLength: 200,
      });
    });
  });

  describe("predefined values", () => {
    it("should be able to set dataset size by clicking button", () => {
      const handleSubmit = jest.fn();
      const wrapper = shallow(
        <Generator
          totalDataset={1200}
          totalLength={100000}
          onSubmit={handleSubmit}
        />
      );
      expect(handleSubmit).not.toHaveBeenCalled();
      wrapper.find("button.preset.dataset").last().simulate("click");
      wrapper.find("Form").simulate("submit");

      expect(handleSubmit).toHaveBeenCalled();
      expect(handleSubmit).toHaveBeenCalledWith({
        dataSetSize: 1200,
        playlistLength: 3600,
      });
    });

    it("should be able to set playlist length by clicking button", () => {
      const handleSubmit = jest.fn();
      const wrapper = shallow(
        <Generator
          totalDataset={1200}
          totalLength={100000}
          onSubmit={handleSubmit}
        />
      );
      expect(handleSubmit).not.toHaveBeenCalled();
      wrapper.find("button.preset.length").last().simulate("click");
      wrapper.find("Form").simulate("submit");

      expect(handleSubmit).toHaveBeenCalled();
      expect(handleSubmit).toHaveBeenCalledWith({
        dataSetSize: 20,
        playlistLength: 100000,
      });
    });
  });
});
