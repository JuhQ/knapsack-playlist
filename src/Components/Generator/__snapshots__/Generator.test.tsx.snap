// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Generator should render 1`] = `
<div
  className="ui container"
>
  <form
    className="ui form"
    onSubmit={[Function]}
  >
    <div>
      Size of data set:
      <br />
      <small>
        This is the amount of random items to be selected from the database for the knapsack algorithm
      </small>
      <br />
      <div
        className="ui large right labeled input"
      >
        <input
          onChange={[Function]}
          placeholder="Enter desired data set size"
          type="number"
          value={20}
        />
        <div
          className="ui basic label label"
          onClick={[Function]}
        >
          videos
        </div>
      </div>
    </div>
    <div>
      Length of playlist:
      <br />
      <div
        className="ui large right labeled input"
      >
        <input
          onChange={[Function]}
          placeholder="Enter desired playlist length"
          type="number"
          value={3600}
        />
        <div
          className="ui basic label label"
          onClick={[Function]}
        >
          seconds
        </div>
      </div>
    </div>
    <button
      type="submit"
    >
      Submit
    </button>
  </form>
</div>
`;
